### DISCUSSION TITLE
Loading TrayIcon menu faster?

### DISCUSSION THREAD

**Discussion by Mtax-Development** [AvaloniaUI/Avalonia#17377] [Q&A]
I am developing an application, which is operated through the `TrayIcon` menu only. On Windows, the first opening of said menu, after right-clicking it, is always very slow. It often takes at least two seconds before the menu is shown, even though all logic of constructing the menu is done when the application first starts, not when the menu is opened. There even are cases where I can even hear my laptop speed-up fans only to load said menu, which then will appear after a delay. However, subsequent menu openings are quick as expected.

Is there a way to preload resources needed to render the menu without user interaction? I attempted to create a window with text off-screen to force text-related resources to load, but I am still running into this issue. Perhaps there is a way to preload DLLs involved in rendering the `TrayIcon` menu and keep them in memory?

**Comment by stevemonaco** [comment]
You should try to narrow down the cause, if possible. If you can't (and suspect long JIT times), then try ReadyToRun deployment. NativeAOT would work, too, but can be hard to port already existing apps to it.

**Reply by Mtax-Development** [reply]
The issue happens under ReadyToRun, as that was one of the first things I tried. I assume introducing Ahead-of-Time compilation is not just done with marking it to be used, as doing causes my application to silently crash on boot.

**Reply by stevemonaco** [reply]
No, NativeAOT has strict requirements on what kind of .NET code you run. Being careful to avoid trimming-related issues and IL codegen are crucial. Many libraries rely on these, especially for convenience. Some have sourcegen ports.

Either way, a 2+s start time seems unusual unless it's an especially low performance device or you're creating a page containing many, many controls. ie. Showing a large collection (500+) items without using virtualization.

**Reply by Mtax-Development** [reply]
The application is fairly simple, about 20 entries in the `TrayIcon` menu. The `TrayIcon` is initialized entirely during the runtime, at application boot. I am not testing on a slow device, but the delay tends to increase when the system is busy, as if my process had low priority overall. Publish is set to produce a single file, but I noticed the DLL the application is loading when this delay happens, `libHarfBuzzSharp.dll`, always remains a separate file. But the delay occurs even if not using the single file setting.

**Reply by timunie** [reply]
Try with a minimum sample and file it here if issue persists

**Reply by Mtax-Development** [reply]
@timunie The project is open-source and can be checked here: https://github.com/Mtax-Development/GameMakerCompanion
Since the application is rather simple, this project in itself is of minimal state of my setup, because the `Tray` class initializes everything else. I would not have much to disable, other than an additional thread started by it, but from my testing, it does not affect the issue.

**Reply by timunie** [reply]
No, this is not a minimal sample, this is a complete App. I would have to scan the entire source before to try and I at least have no capacity to do this.

**Reply by Mtax-Development** [reply]
@timunie, apologies for the hiatus. I prepared a minimal sample presenting the issue, where I trimmed the project to minimum and included a `.gif` file visually presenting the issue. The delay shown in it is proven by that the cursor can be moved after performing a right-click on the Tray Icon to cause the Menu to appear in a different spot, other than right above of the icon. This is exemplified best upon the **first** right-click of the runtime, so keep restarting the application to test it multiple times. 

[Tray Menu Opening Delay Minimal Sample.zip](https://github.com/user-attachments/files/18209919/Tray.Menu.Opening.Delay.Minimal.Sample.zip)

**Reply by Mtax-Development** [reply]
Any chance I could get a reply on this? For now, I have exhausted all options I could do before knowing whether or not this is a quirk of Avalonia or something that can be affected the side of a depending project.

**Reply by timunie** [reply]
Maybe it's a bit quicker if you move it into `public override void OnFrameworkInitializationCompleted()`

Other than that I've no clue why it is quicker if I define it in App.xaml 🤷

**Reply by stevemonaco** [reply]
Really strange repro. You stripped away the debug configuration so the app can't be debugged outside of "printf debugging".

Anyways, I can reproduce the effect in your app: long first-time show (around 1s) and instant afterwards. In a simple standalone app (with an Avalonia window), this feels quicker (a few hundred ms, then instant).

I've migrated your app into a more typical workflow for a tray-only app and experience similar delays, so I assume there are some lazy initialization costs not being paid in this scenario until the tray menu is opened. Maybe initializing theme resources.

I can't figure out a way to force this initialization without showing a visible window or such.

**Reply by Mtax-Development** [reply]
I performed normal-usage testing throughout the week and it does not seem that delays behave different when Tray Menu initialization is moved to `OnFrameworkInitializationCompleted()`.

So, what happens from this point? Is this issue treated as a bug in Avalonia now? Can I do anything else?

**Reply by Mtax-Development** [reply]
@timunie, any chance for a reply?

**Reply by timunie** [reply]
Ah sorry, had you on to-do but wasn't there yet. I have currently no idea how to improve that and I can also say I can reproduce here. Feel free to open a new issue with the repo linked / attached for further investigations. However I don't see any fixes in the near future. 

> [!NOTE]
> If you have paid support, you can ask for priority.
> If you have an idea, add your test into `src > Samples > Sandbox` and file a PR. 

Sorry for not having a simple fix at this time.
